src/ck-core/ckcheckpoint.C

```diff
@@ -214,14 +220,25 @@ void CkCheckpointMgr::Checkpoint(const char *dirname, CkCallback cb, bool _reque
 	requestStatus = _requestStatus;
 	// make dir on all PEs in case it is a local directory
 	CmiMkdir(dirname);
-	bool success = true;
-        if (CmiNumPartitions() > 1) {
-          ostringstream partDir;
-          partDir << dirname;
-          addPartitionDirectory(partDir);
-          CmiMkdir(partDir.str().c_str());
-        }
 
+  // Create partition directories (if applicable)
+  ostringstream dirPath;
+  dirPath << dirname;
+  if (CmiNumPartitions() > 1) {
+    addPartitionDirectory(dirPath);
+    CmiMkdir(dirPath.str().c_str());
+  }
+
+  // Create subdirectories by grouping PEs
+  // Nodegroups are fine since # of logical nodes <= # of PEs
+  int numPes = CkNumPes();
+  int numPeGroups = ceil((double)numPes / PE_GROUP_SIZE);
+  for (int i = 0; i < numPeGroups; i++) {
```

Comment:
Do we want all of the PEs to create all of the directories? The partitions code only creates the directory for the current partition.

Reply:
there may be a race condition if you assume another pe has created the folder... but can be fixed by looping on error condition by non-creator PEs.  Right?

Reply:
It should be fine to have a race with `CmiMkdir`, calls after the directory is created shouldn't do anything. There is similar race already with `CmiMkdir(dirname)` in line 222. I'll change the code to have each PE try to create a folder only for its PE group (PEs 0-255 try to create folder for PE group 0, PEs 256-511 for PE group 1, etc.). I think it's necessary to have 256 PEs try to create a folder because PEs in a group might be on different physical nodes.